### PR TITLE
[Android] Fix BasicMessageChannel.resizeChannelBuffer

### PR DESCRIPTION
**Pull request by bleroux**
## Description

This PR updates `BasicMessageChannel.resizeChannelBuffer` implementation. Previous implementation builds a wrong `ByteBuffer` and was not tested so it is difficult to know when it stopped working.  

## Related Issue

Fixes https://github.com/flutter/flutter/issues/126632

## Tests

Adds 1 test.

### CODE REVIEW

**Comment by gmackall** [comment]
Note from android triage: @gaaclarke Do you know who might be a good reviewer for this?

**Comment by chinmaygarde** [comment]
I don't think this is a fix. We are just swallowing an exception.

File: shell/platform/android/test/io/flutter/plugin/common/MethodChannelTest.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+
+    ByteBuffer packet = null;
+    try {
+      final String content = String.format("resize\r%s\r%d", "flutter/test", 3);
+      final byte[] bytes = content.getBytes("UTF-8");
+      packet = ByteBuffer.allocateDirect(bytes.length);
+      packet.put(bytes);
+    } catch (UnsupportedEncodingException e) {
+      throw new AssertionError("UTF-8 only");
+    }
```
Comment (dnfield) [inline shell/platform/android/test/io/flutter/plugin/common/MethodChannelTest.java:42]: Why is this code in the test? It's not actually contributing to the test.

File: shell/platform/android/io/flutter/plugin/common/BasicMessageChannel.java [hunk 2]
```diff
@@ -145,10 +144,15 @@ public void resizeChannelBuffer(int newSize) {
...
-    messenger.send(CHANNEL_BUFFERS_CHANNEL, message);
+    try {
+      final String content = String.format("resize\r%s\r%d", channel, newSize);
+      final byte[] bytes = content.getBytes("UTF-8");
+      ByteBuffer packet = ByteBuffer.allocateDirect(bytes.length);
+      packet.put(bytes);
+      messenger.send(BasicMessageChannel.CHANNEL_BUFFERS_CHANNEL, packet);
+    } catch (UnsupportedEncodingException e) {
+      Log.e(TAG, "Failed to resize channel buffer named " + channel, e);
+    }
```
Comment (dnfield) [inline shell/platform/android/io/flutter/plugin/common/BasicMessageChannel.java:155]: This seems to be unrelated to the bug filed upstream and intended to support channel names that aren't valid as UTF-8 strings or something. Can it just be removed or perhaps serpated out into its own PR?

Comment (jason-simmons) [reply shell/platform/android/io/flutter/plugin/common/BasicMessageChannel.java:155]: The `catch` is necessary because `String.getBytes(charset)` is declared as `throws UnsupportedEncodingException`

(In practice the exception will not be thrown because UTF-8 is guaranteed to be supported)

Alternatively, this could continue using `Charset.forName` (which does not throw any checked exceptions)

File: shell/platform/android/test/io/flutter/plugin/common/MethodChannelTest.java (hunk 1 above)
Comment (bleroux) [reply shell/platform/android/test/io/flutter/plugin/common/MethodChannelTest.java:42]: The 'packet' variable is used line 47.

**Comment by jason-simmons** [comment]
> I don't think this is a fix. We are just swallowing an exception.

The fix here is the use of `ByteBuffer.allocateDirect` instead of `ByteBuffer.wrap`.

(The current Android Java/JNI platform message implementation assumes that all buffers passed to native are direct buffers)

**Comment by bleroux** [comment]
> I don't think this is a fix. We are just swallowing an exception.

@chinmaygarde The fix is not to swallow an exception is it to replace the way the `ByteBuffer` is build.
Previously it relied on `ByteBuffer.wrap` and throws a `RangeError`, see https://github.com/flutter/flutter/issues/126632#issue-1706950737.

The fix is to rely on `ByteBuffer.allocateDirect`.

File: shell/platform/android/test/io/flutter/plugin/common/MethodChannelTest.java (hunk 1 above)
Comment (dnfield) [reply shell/platform/android/test/io/flutter/plugin/common/MethodChannelTest.java:42]: Ahh I see now.

**Comment by bleroux** [comment]
I have updated the PR to use `Charset.forName` as suggested by Jason and it makes the PR simpler and less confusing. Thanks @jason-simmons!

**Review by jason-simmons — APPROVED** [review]

**Review by dnfield — APPROVED** [review]

**Comment by Hixie** [comment]
FWIW, the format used for the message in this PR is deprecated, ideally we would use the binary format instead. It should be slightly more efficient.
